### ISSUE TITLE
Testing Bash Pipelines + Metrics

### ISSUE THREAD

**Issue by rennergade** [Lind-Project/lind_project#56]
Now that bash is running reasonably we'd like to test it on a number of pipelines and compare our metrics to native bash runs.

To accomplish this we should:

1. Create a list of typical bash pipeline commands ie `ls | cat` or `cat file | grep -v a | sort - r`.
2. Create a script that runs these commands natively and with Lind, and stores runtime metrics
3. Use this script to collect data we can analyze

@JustinCappos How many pipeline commands do you think we need to collect a reasonable data set?

**Comment by JustinCappos** [comment]
Right now, just getting *any* data would be helpful.  How much we need
depends on what we end up wanting to show, etc.  So, let's do a little
exploration then we can scope it better.

**Comment by rennergade** [comment]
I outlined what we want a testing program to generate data for these pipelines to look like.

This should eventually just let us supply a textfile full of pipelines which will run bash through Lind and natively, collect data, and then generate plots.

https://docs.google.com/document/d/1ljg6mWFKSLo7ityxrOKMytKZLr8Vye1MvwlCrGIV2Ig/edit#

**Comment by rennergade** [comment]
I was able to download the source code for grep and build it in the same way we were building coreutils. I used `grep` in some test scripts and it seems to work fine when compiled with nacl.

**Comment by rennergade** [comment]
The `sort` coreutil file is not working in bash scripts. I get the following error:

``executing: [sel_ldr -a -- "/lib/glibc/runnable-ld.so" --library-path "/lib/glibc" /bin/bash catgrepsort]
Lind v1.0-rc2
Opening file system...
done.
[14789,1890141248:11:23:06.982167] BYPASSING ALL ACL CHECKS
/bin/sort: couldn't create temporary file: /tmp/sortEW0M8V: Invalid argument
```

This probably occurs in anything that needs a temp file, will create a separate issue.

**Comment by rennergade** [comment]
I'm still working on setting up the testing suite but ran some initial results manually:

`catgrepwc` which cats a file, greps it for a string, and gets the wc of that grep:
Lind - 0.658 s
Native - 0.010 s
Native ~66x faster


`lscat` which calls ls, and cats the output:
Lind - 0.493 s
Native - 0.008 s
Native ~61x faster

`lscat` which calls ls, and cats the output to wc:
Lind - 0.660 s
Native - 0.009 s
Native ~73x faster



`

**Comment by rennergade** [comment]
I've added some testing capabilities into SafePOSIX to get runtimes for each time a "system call" is made.

Preliminary results show that any call to `persist_metadata` is taking around 35ms. These are mostly found when the `access` system call is made. In total, this is adding around ~20% overhead to Lind runtimes.

Other system calls that are taking longer than average are file calls such as open, close, write, and read, as well as the stat system call.

**Comment by JustinCappos** [comment]
Can you look at what a security layer says about the times taken by readat,
writeat, openfile, etc.?  I imagine this is a substantial part of the whole
execution time.

Justin

**Comment by rennergade** [comment]
Justin had asked above about the times of our emulated write/read/openfile calls hitting the system. My tests are showing that these are taking less than a tenth of a millisecond each, which seems really good.

[I made this spreadsheet](https://docs.google.com/spreadsheets/d/1alMfy5U_l774pmAxCf9HX81jknbsDXvoRmaT0Na7PiQ/edit#gid=0) to outline the specific system call times for a run using the `ls | cat | wc` pipeline.  The total lind runtime is exaggerated a bit (+ ~100ms) because of the frequent prints for timing information, but this still gives us an idea of whats happening proportionally.

The most obvious things are:

1) `access()` is taking a ton of time. I've tracked this slowness down to the `serializedata` function which is recursively serializing the data in persist_metadata

2) The time contributed from sources other than SafePOSIX (NaCl and the usermode) is quite large. I think it's pretty safe to assume that these contributions are overwhelmingly from NaCl.

**Comment by JustinCappos** [comment]
Interesting.  I'm a little surprised that serialization is so slow, but I
guess this was something we wrote in Python and never tried to optimize.

**Comment by rennergade** [comment]
Profiling NaCl with `prof` on a small-file bash pipeline seems to indicate that the vast majority of NaCl related overhead is due to Execution-Context copying on fork() calls.

![perf-kernel](https://user-images.githubusercontent.com/7719272/81850774-72783d80-9526-11ea-8a6a-3c5c53696dd6.png)

**Comment by rennergade** [comment]
I've started testing pipelines with large files. My initial test program involves using the text for the King James Bible (>4MB) and catting that to word count `cat bible.txt | wc`.

Native run is about 20ms. Lind runtime is ~ 2 million ms. Not great!

What I was first seeing here while this was slowly running was a bunch of slow ~20 second reads. But after awhile I would see some huge 250-400 second write system calls.

Both are probably inefficient, but my inclination for now is that read is mostly waiting for write. My implementation for pipe writes is to write the bytes char by char to the "pipe" list. O(n) for this many bytes is seemingly very very slow.

**Comment by rennergade** [comment]
Haven't updated this in awhile since I've been communicating on multiple platforms.

I implemented a much faster list method that did bulk append and remove, after that I was able to make that even faster by using a deque and using itertools.

However these still aren't comparable to native, and I'm investigating why.

At Justin's suggestion I ran some baseline tests, and collected system call times (but don't print the times until after program run to not majorly effect results). To do a baseline I basically replaced the pipe write function with writing nothing, and had read just return count * 'A'.

This was taking surprisingly long even for a long file. In fact, the sum of my system call times was 2x as long as what lind reported as the final user time. This was a headscratcher. I then realized that the implementation was basically having timers run simultaneously while operations hit locks, double the timers double the time! And this was happening a lot, basically the entire program time by percentage!

Looking at the logs, it seemed that even though I removed the write, it was still attempting to write all of those bytes, and NaCl limits the buffer size tremendously. It looks like in NaCl write buffers are limited to 2^17 while read buffers are limited to 2^14, these are both 2^32 in native linux.

So it kept on trying to write blank data until NaCl thought it wrote the entire file, thus no EOF for awhile, at the same time read, though it only read data once was waiting on that EOF and continuously blocking waiting for writes.

**Some basline test numbers - all of these tests are done with a 40MB super-bible file except first**
1) Short program (ls | cat) with same number of pipes: 420 ms
2) Pipe of bible to wc (cat largebib.txt | wc) but write writes nothing, and first read returns all A's, subsequent reads return 0 - 643 ms
3) """ but all reads return A's, for same size as bible file - 1417 ms
4) """ but with current deque pipe implementation - 4449 ms

**Comment by rennergade** [comment]
**Some general system call metrics:**

I mad a testing program that does file operations: open/write/seek/read/close with a simple 256 char buffer, here are some lind/native comparisons

**Open**
Native - 23 us
Lind - 1739 us

**Write**
Native - 3 us
Lind - 220 us

**Seek**
Native - 4 us
Lind - 43 us

**Read**
Native - 3 us
Lind - 193 us

**Close**
Native - 3 us
Lind - 81 us

**Comment by rennergade** [comment]
Getting back to this after a several week hiatus - some of the off time was good to cleanse my thought process here.

I think my efforts to finely track problem areas by inserting into the code was somewhat misguided and hard to decipher. My thoughts at this point is trying to use `perf` as much as possible is the desired route. I've scripted a way to generate a bunch of flamegraphs on various bash pipelines, I have like 10-12 pipelines to test now but should increase that further. 

What I need to figure out is how to look at the results with a finer grain. Hopefully I can talk to Brendan soon about optimizing perf. It also would be great to get similar results for the Python portions, because our results right now aren't very specific. Santiago suggested [this tool](https://github.com/evanhempel/python-flamegraph) but I need to learn more about Repy to figure out how to hook this in/or if it's possible. Hoping to pick Justin's brain on this soon as well.

In general, my feeling is we need more tests at a finer grain, rather than trying to manually pick this apart.

**Comment by rennergade** [comment]
Per discussion with Brendan:

1) Use GDB to get backtraces of interesting function calls in large "catcatcatlbwc" flamegraphs, and see what is calling them between sel_ldr.

2) Keep on trying to get Python Flamegraphs, need to figure out how to patch into RePy. Grepping for `rexec` might reveal what restricted execution is preventing this.

3) Add some more long tests so we have more to look at.

**Comment by rennergade** [comment]
I did backtraces of the `sem_post@@GLIBC` and `do_futex_wait.constprop.0` functions that  looked interesting from the long FlameGraph, these we're both taking significant chunks of time. Interesting enough they seem to be related. Sem_post seems to be from Python lock releases, while do_futex_wait seems to be from python lock acquiring.

Logs of sample backtraces below:

[do_futex_wait.log](https://github.com/Lind-Project/lind_project/files/4996403/do_futex_wait.log)
[sem_post_backtraces.log](https://github.com/Lind-Project/lind_project/files/4996404/sem_post_backtraces.log)

**Comment by rennergade** [comment]
I ran some tests and found out some interesting things about the above functions.

**sem_post**

The `sem_post` function is basically always happening once we get into Python at any time. It happens during the init and finalize function calls, which don't even have explicit locks, as well as with all the other calls.

I also ran gdb on regular ol' python2, running a helllo world script and then broke for sem_post. There were tons of those happening all the time just on a simple unthreaded program.

**do_futex_wait**

On the other hand, this function is only happening in specific cases. I saw none of these in a basic python debug, and this doesn't even occur on a single threaded Lind program. We only start seeing these after a lind_fork call.

I went back to basic python and inputted a sample multithread script in gdb. Then I started seeing `do_futex_wait` as well here.

I think the take away here is that this isn't our in-program locks but is a Python GIL thing. 

The question now is:
1) This something we're just going to have to deal with
2) Should we be patching this to reduce kernel calls some way, is this possible?
3) Is the overhead here not because of the GIL itself but just a bad pipe implementation.

**Comment by rennergade** [comment]
I ran a pure python2-multithread test where I had the threads run for a substantial time and took a flamegraph of it. The results are below.

[flame-graph of python multi-thread](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/python-mthread.svg)

As far as the do_futex_wait and sem_post stacks, this looks a lot like the stacks we saw in the Lind pipeline flamegraph.

I think that rules out my question for (3) above, meaning this overhead is pure GIL and doesn't arise more because of the pipe implementation.

**Comment by rennergade** [comment]
I brought back the separate pipe-testing python file I had made earlier and used it to create a flamegraph. This time I created an even bigger file ~600 MB and ran that through the test script (code below) as well through lind and natively.

Natively piping this file took ~2.5s
The test script takes about 15 seconds to "pipe" this file
Lind takes ~ 86 seconds

[Flamegraph for Test](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/python-deque.svg)
[Flamegraph for Lind](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/lind-deque-mb.svg)


```
import sys
import time
import threading
from collections import deque
from itertools import islice, starmap, repeat

WRITE_BUFF_SIZE = 131072

class emulated_lock (object):

    __slots__ = ["lock"]

    def __init__(self):
        # Create our lock
        self.lock = threading.Lock()


    def acquire(self, blocking):
        return self.lock.acquire(blocking)


    def release(self):
        try:
            self.lock.release()
        except thread.error:
            print "Releasing an un-locked lock!"




def _read_from_pipe(fd, count):

    # lets find the pipe number and acquire the readlock
    pipetable[pipenumber]['readlock'].acquire(True)

    data = ""

    # we're going to try to get bytes up until the amount we requested, but break if we there's nothing there and we get an EOF
    while True:
      try:
        current_pipesize = len(pipetable[pipenumber]['data'])

        if current_pipesize == 0 and pipetable[pipenumber]['eof'] == True:
            break


          # If count is smaller than pipe, read that much and delete it from pipe,
          # if not, take the whole thing
        if count < current_pipesize:  
            data += "".join(starmap(pipetable[pipenumber]['data'].popleft, repeat((), count)))
            break
        else:
            data += "".join(starmap(pipetable[pipenumber]['data'].popleft, repeat((), current_pipesize)))
            count -= current_pipesize

      except IndexError, e:
          continue

    # release our readlock  
    pipetable[pipenumber]['readlock'].release()
    return data


    # helper function for pipe writes
def _write_to_pipe(fd, data):

    # find pipe number, and grab lock
    pipetable[pipenumber]['writelock'].acquire(True)

    # append data to pipe list

    pipetable[pipenumber]['data'].extend(data)

    # release our write lock     
    pipetable[pipenumber]['writelock'].release()

    return len(data)

def writeWrapper(fd, data):

    byteswritten = 0

    datalen = len(data)
    datalist = list()
    buffersize = 2**14
    iterations = (count/buffersize) + 1

    for i in range(iterations):
        datalist.append(data[(i*buffersize):((i+1)*buffersize)])
        


    timestart = time.time()
    for i in range(iterations):
        byteswritten += _write_to_pipe(fd, datalist[i])
    timeend = time.time()

    wtime = (timeend - timestart)
    print "Wrote " + str(byteswritten) + " bytes in " + str(wtime) + " seconds"
    pipetable[pipenumber]['eof'] = True

def readWrapper(fd, count):

    buffersize = 2**14

    iterations = (count/buffersize) + 1

    data = ''

    timestart = time.time()
    for i in range(iterations):
        buffer = _read_from_pipe(fd, buffersize)
        data += buffer
    timeend = time.time()


    rtime = (timeend - timestart)

    print "Read " + str(len(data)) + " bytes in " + str(rtime) + " seconds"


pipetable = {}
pipenumber = 1
pipetable[pipenumber] = {'data':deque(), 'eof':False, 'writelock':emulated_lock(), 'readlock':emulated_lock()}

readfd = 0
writefd = 1

with open (sys.argv[1], "r") as bible:
    bibletext=bible.read()

count = len(bibletext)
print "Filesize " + str(count) + " bytes"

writeThread = threading.Thread(target=writeWrapper, args=(1, bibletext))
readThread = threading.Thread(target=readWrapper, args=(0, count))

writeThread.start()
readThread.start()

writeThread.join()
readThread.join()

print "done"
```

**Comment by rennergade** [comment]
I wanted to document some earlier results from the `./write_to_pipe | ./read_from_pipe` example. This uses a timer in each of the programs wrapped around the write and read system calls. The testing sends a 2^14 byte buffer 2^11 times.

I ran this natively and through Lind. And then modified SafePosix so it returned at increasingly earlier stages of the write (in the fs_read stub, in the dispatcher, in the initial SafePosix reception, before sending to SafePosix in the NaCl RPC stage.

**Times for write()**
Native - 7 ms
Lind - 752 ms

Lind with Pipe doing a "blank" write - 144 ms
Blank "write", removing writelock - 119ms
Returning from fs_write.repy without entering fs_calls - 38ms
Returning from dispatcher = 33 ms
Returning after first SafePosix interception - 7 ms
Returning in NaCl before creating RPC message - 0ms

**Comment by rennergade** [comment]
I decided to run the above test the other way. Instead keeping the write steady but mocking read.

By returning `count * 'A'` instead of reading from the pipe, that immediately reduced the Lind read time to 0ms.

**Times for read()**
Native - 7 ms
Lind - 475 ms
Lind, pipes modified to return constant buffer - 0 ms

**Comment by rennergade** [comment]
To Do list for tests this week:

- Lind pipe baseline with write/read turned off in NaCl
- SafePosix "chop" but with smaller writes/reads

**Comment by rennergade** [comment]
Answer for question: Investigate more into buffer sizes (write/read) - how are they broken up? Is this program specific (cat) or general?

This isn't Lind/NaCl specific. It's being broken up by `cat`. This is explained [here](https://unix.stackexchange.com/questions/245499/how-does-cat-know-the-optimum-block-size-to-use) and I confirmed this on our platform. When using cat we have the same buffersizes defined from IO_BUFSIZE.

**Comment by rennergade** [comment]
Doing some basic timing of python string functions suggests that these aren't the major timing roadblock. Even the most inefficient copying for large buffers (2^17 like we've been seeing) takes a fraction of a millisecond. 

I wanted to just get times for the deque implementation without dealing with the threading to see how long each takes minimally without dealing with the GIL.

Results for 4047392 bytes
Write: 52.8 ms
Read: 91.4 ms

For reference a `cat file | cat` of the same size takes 20 ms. This obviously isn't apples to oranges because bash forks these and runs them in parallel.

**Comment by rennergade** [comment]
I spent the last couple of days looking at and getting data to better profile system call times for long pipelines with pipes turned "on" or "off". I also figured out how to hack SafePosix to give us data on what kind of write/read is happening, which gave some interesting data on what is actually slow for the reads.

You can see this laid out here: https://docs.google.com/spreadsheets/d/1dcHhKhtvjPG_zySUpP9H5cgcx2k34lUEvBANDhVvRWY/edit#gid=2025765307

Included is full system call logging for cat'ting a 600MB file to wc, which includes system call name, whats happening if its a write or read, and the time. This is done both with pipes turned on, or with them being cut off and having their data faked. Also included is data from strace for this same bash pipeline run natively, using the -fc options to log average times over the forked program.

The biggest thing that jumps out is that the pipes aren't the slowest thing here, its the file reads from cat. Justin has been talking about Nanny and this very likely could be from that. Most of these are like ~12 ms each where the pipe reads even with pipes on are ~0.25 ms.

The other thing that jumps out is that system calls are still pretty slow compared to native. The times from `strace` are mostly in the single digit microsecond range where our calls are mostly in the tens of microseconds and reads/writes in the hundreds of microseconds.

I think this is a decent mapping of whats going on here so will be good to discuss this.

**Comment by rennergade** [comment]
Disabling Nanny made this more than 3x faster. I've added those results to [the spreadsheet](https://docs.google.com/spreadsheets/d/1dcHhKhtvjPG_zySUpP9H5cgcx2k34lUEvBANDhVvRWY/edit#gid=2077766252).

Something weird happened here. In previous runs, the totaled syscall time was more than the program time (system calls are running in parallel during the run of the program, thus higher times). But once we reduced Nanny, the program time is significantly (in this case 10 seconds) less than the total of all syscall times (still running in parallel). For reference, the native run for this pipe test was clocked at 2.2 seconds.

**Comment by rennergade** [comment]
I added a FlameGraph for the run with nanny now disabled as well [here](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/catmbwc-nonanny.svg)

Some great news, getting rid of Nanny basically got rid of the futex/sempost stacks that were very present before. Now we can clearly see some moderate sized chunks related to the deque implementation (collections.so, string_join, iter_next, extend).

On the other hand, there's this big chunk in the middle that gives us very little clue to what it is. The one clue is that the "fork" stack (on the right) is sitting above it, this leads me to believe it's NaCl related.

**Comment by JustinCappos** [comment]
Please also disable the CPU monitor for repy (in nanny) if it isn't already
done.

**Comment by rennergade** [comment]
ToDo - From talking with Justin and Brendan today:

- Run perf with higher frequency, run perf on version with pipes off
- Look into supplying perf symbols
- Run program comparing basic system calls to strace results
- Go back to hacking with read and write pipe programs

**Comment by rennergade** [comment]
I was able to generate some much clearer FlameGraphs by adding a flag for to deal with dwarf formatting. [Here is the new example for the 600 MB pipe.](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/catmbwc-highfreq-dw.svg)

As you can see it's much clearer in the stacks of how Lind is making function calls, and most of the program is being taken up by writes/reads as we can expect. There still is large unknown piece on the right which I'm still confused about, as well as a small section labeled [anon] on the left that includes some python calls that should be on the main program stacks. Regardless this looks much better and is much more helpful. It looks here like as far as the reads/writes go the tools used for the pipe implementation are still the main time contributors.

**Comment by rennergade** [comment]
I wanted to test just doing a large write to a file instead of a pipe in terms of testing here, because I hadn't really done that before. I got some encouraging results. I redirected output here to a file because of the the time associated with writing to the terminal and some problems timing terminal output in Lind.

Both programs writing 4GB to stdout, both redirected to a file:
Native: 10.8 seconds
Lind: 8.48 seconds

This is the first time we've seen Lind outperform native, which is pretty great.

**Comment by rennergade** [comment]
I wanted to get some times for a simple program to time basic system call. This test file does open (with pre-created file), writes, seeks, reads, and closes. Some times using different methods:

**Lind (timer in program)**
open usec: 218 us
write usec: 54 us
seek usec: 58 us
read usec: 88 us
close usec: 59 us

**Native (timer in program)**
open usec: 24 us
write usec: 19 us
seek usec: 3 us
read usec: 8 us
close usec: 5 us

**Native (strace -T)**
open  32 us
write 35 us
lseek 20 us
read 30 us
close 22 us

**Lind - SafePOSIX interception timer**
open 107 us
write 60 us
lseek 34 us
read 73 us
close 56 usec

**Comment by rennergade** [comment]
Above I reported:

```Both programs writing 4GB to stdout, both redirected to a file:
Native: 10.8 seconds
Lind: 8.48 seconds```

However this was using `lind ./lotsofwrites.nexe &> log` which had lind writing to stdout via `write` and then native bash handling the redirect into the file. I wanted to make it so lind was writing to the file, so uploaded a script that did `lind ./lotsofwrites.nexe &> log` and ran that in Lind through `lind /bin/bash script`

This took 3800 seconds, almost a 500x slowdown from letting native bash handle the redirect, and this is running the version with nanny disabled. However, upon running a second time it took only 26 seconds, which indicates that creating and persisting huge newfiles in lindfs takes an incredibly long time. I confirmed this by deleting the file it wrote too and trying again, and it went back to taking a very long time. 

I'm going to upload a flamegraph in a bit.

**Comment by rennergade** [comment]
At Justin's suggestion I'm going back to the read/write pipe programs run through a bash script. This entails a program that writes 16KB blocks n number of times for at total of N bytes, and a program that reads until EOF is received. Both the write portion and read portion of each program have timer's wrapped around them.

I did this for several total write sizes: 2MB, 32MB, 64MB, and 256MB and logged times as well as generated FlameGraphs.

**2mb**
[FlameGraph](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/rwpipe2mb.svg)
Native:
Total - 252ms
Write - 0ms
Read - 0ms

Lind:
Total - 670ms
Write - 49ms
Read - 31m

Note: This is before our fork improvement is added so this is still Fork dominated.

**32mb**
[FlameGraph](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/rwpipe32mb.svg)
Native:
Total - 260 ms 
Write - 7 ms
Read - 7 ms

Lind:
Total - 1547 ms
Write - 773ms
Read - 479m

Note: Fork and Init are still contributing here but you begin to see the read and writes


**64mb**
[FlameGraph](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/rwpipe64mb.svg)
Native:
Total - 269 ms 
Write - 14 ms
Read - 14 ms

Lind:
Total  - 83908 ms
Write - 83484 ms
Read - 82662 ms

Note: Write and read now dominate, with read growing much faster (from flamegraph)

**256mb**
[FlameGraph](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/rwpipe256mb.svg)
Native:
Total - 269 ms 
Write - 14 ms
Read - 14 ms

Lind:
Total - 831753 ms
Write - 831342 ms
Read - 830617 ms

Note: Even more pronounced than 64mb

**Comment by rennergade** [comment]
I did another test running basic file calls through Lind and Native. I ran the following program (snippet below) which does a bunch of file calls 10,000 times.

```
    for (int j = 0; j < 100000; j++) {

    	int test_fd = open("filetimes.txt", O_RDWR);
    	write(test_fd, buffer, WRITE_BUFFER_SIZE);
    	lseek(test_fd, 0, SEEK_SET);
    	read(test_fd, readbuffer, WRITE_BUFFER_SIZE);
    	close(test_fd);
    }

```

Native took 145 ms while Lind took 36731 ms, a very substantial 253x slowdown.

[The flamegraph for Lind is here](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/fileloop.svg)

Here are the averages in us for native:

open | 2
fstat | 9 (only called 3 times)
stat | 16 (only called 16 times)
write | 2
read | 1
lseek | 1
close | 1


As opposed to Lind

open | 65
fstat | 58 (happens all 10,000 times)
stat | 69 (happens all 10,000 times)
lseek | 30
write | 47
read | 38
close | 49

**Comment by rennergade** [comment]
Talking with Brendan today, it seems our conclusion is that we still need to know a lot more about whats going on in SafePOSIX but our current tooling is inadequate. 

We need to be able to capture framestacks and times from Python. Hacking perf/python seem to be the best options. It may be worth revisiting PyFlame and seeing why I wasn't getting the results that they advertised and hacking that to work with Lind.

**Comment by rennergade** [comment]
After discussing the file loop test (referenced two posts above this) we were pretty stunned about the massive time discrepancy between Lind and Native. I checked some things and ran it again, and I can confirm that there is a massive slowdown.

What's strange is that looking through the logs for individual system calls, none of them are taking more than 0.1 ms. It's looping 10,000 times, and doing 7 system calls each time open (with a stat and fstat each time, write, seek, read, close.

Back of the envelope thats 7 x 10,000 x 0.1 = 7,000 ms which intuitively makes much more sense. I'm not sure where the rest of the overhead is coming from.

**Comment by rennergade** [comment]
I went through the log and actually combed through each call. There are actually 152 calls over 0.1 ms of the total 70,000. Of these, 138 are less than 0.2 ms, one is slightly less that 0.5 ms, 7 hover around 1 ms, and then we have a stray 2, 4, 6, 8, 10 and 45. Most of the slightly longer ones are opens with the outliers being reads/writes.

Even adding these extra calls in to the envelope calculation, we're hitting at most 10,000 ms here, not the even crazier 36,000 ms. What's also weird is that the way the timers are set up, they're encompassing everything in SafePOSIX, so this extra overhead SHOULD be in native client. However, I'm not seeing that in the FlameGraph.

**Comment by rennergade** [comment]
It seems like I've gotten Python profiling to work through a different tool, [py-spy](https://github.com/benfred/py-spy). Interestingly enough, this uses `process_vm_readv` to directly read memory from the Python instance. Pretty cool! Another fun introduction to the `process_vm` family.

[Here's the first Lind graph I generated](https://raw.githubusercontent.com/Lind-Project/lind_project/pipeline-test/graphs/pyspy-pipe64.svg). There are still some kinks I need to work out. I can only use a really low sampling rate right now without running into errors. The callstacks are also not as deep as I would think they would (ex. `read_from_pipe` uses the `join()` function a ton`). I'm not sure if this is because of the sampling or due to something else. Either way, this is some progress!

**Comment by rennergade** [comment]
Another cool thing about py-spy is that it has a `top` mode where you can track the function calls in real-time. I'm not seeing the library calls in this mode either sadly.

![Screenshot from 2020-09-15 16-33-42](https://user-images.githubusercontent.com/7719272/93261867-4e956480-f771-11ea-95ce-298fcc750ff3.png)

**Comment by rennergade** [comment]
I've hacked up SafePOSIX to give timer results for each "level" across system calls and log results post-runtime. For these results "syscall time" is the time of the full syscall in safeposix upon interception, "dispatcher time" is the time once it reaches the dispatcher, "stub time" is the time once it reaches its individual stub file as in "fs_write.repy", and "fs call time" is the time it takes to do the actual system call implementation that lies in `lind_fs_calls.py`.

I used this on a basic filetest run (the same open/write/seek/read/close example), as well as running strace to get native call times. I've included the full logs for lind and strace below, but have extracted the core system calls from the actual program here (the rest of the syscalls are loading and exiting, but are useful for comparison of those stages)

Below are the isolated times, which seem encouraging as a first step. The total syscall time for each of these is more than native, but the actual implementation times are either faster than native, or less than 1.5x slower. There is however a decent amount of overhead from interception until the actual implementation, with the largest gaps between the interception and dispatcher, and between the stub and the implementation. 

i'll investigate these further and try to provide more results soon.

Syscall open syscall time 79.0 us dispatcher time 69.0 us stub time 66.0 us fs call time 54.0 us (Native Time - 38 us)
Syscall write syscall time 54.0 us dispatcher time 45.0 us stub time 42.0 us fs call time 33.0 us (Native Time - 44 us)
Syscall seek syscall time 34.0 us dispatcher time 25.0 us stub time 21.0 us fs call time 8.99999999998 us (Native Time - 25 us)
Syscall read syscall time 41.0 us dispatcher time 33.0 us stub time 30.0 us fs call time 21.0 us (Native Time - 33 us)
Syscall close syscall time 55.0 us dispatcher time 46.0 us stub time 43.0 us fs call time 36.0 us (Naitve time - 27 us)

[lind-filetest.txt](https://github.com/Lind-Project/lind_project/files/5270145/lind-filetest.txt)
[strace-filetest.txt](https://github.com/Lind-Project/lind_project/files/5270146/strace-filetest.txt)

**Comment by rennergade** [comment]
In the meeting today I talked about how the largest increases in SafePOSIX rpc time were from syscall interception->dispatcher and from stub->call implementation. We talked about how the closure which correlates syscall implementation with cage may be adding overhead for stub->call. 

Adding some extra timers, it looks like that the majority of the interception->dispatcher portion is just from acquiring the lock in LindSyscall. In general, acquiring locks is expensive (~10 us), and currently we're not very efficient about their use. Some of our fixes should be oriented towards efficient lock use, reducing the time it takes to acquire locks, or both.

**Comment by rennergade** [comment]
https://stackoverflow.com/questions/5426533/python-threading-lock-slows-my-app-down-considerably

The above stack overflow thread has some good commentary on threading and lock performance issues.

It seems using Lock() instead of RLock() will speed up performance if suitable. Even moreso it seems that there is a huge increase in thread/lock performance when switching to Python 3. That has been discussed in the past so worth mentioning again.

**Comment by rennergade** [comment]
I was mistaken above, the Repy locks are Lock() and not RLock().

I imported normal threading locks without the overhead from the repy wrapper and ran some tests. Normal locks take about 1 us to lock/unlock, however Repy seems to make that process 8-16x slower.

**Comment by rennergade** [comment]
Brendan last week suggested running just a simple getpid() test.

Natively, a program that just runs getpid takes 1-2 ms, with the total syscall time being between 0.3 ms and 1.3ms. The actual getpid syscall takes ~11 us.

Running in Lind, we see a total runtime of 22 ms, with the total repy time at 3.9ms, and the pure syscall time at 1.3ms. The actual getpid syscall takes 0 us!

This suggests that 1) Lind syscalls implementations can be very fast! 2) Reiterates that the actual system call implementations are performing at a similar level to native 3) There is still a decent overhead from NaCl in short programs.

The pure syscall

**Comment by rennergade** [comment]
I re-ran the filetest and the filetest with looping (10k times) with the new timing system.

What's interesting is that though the syscall times are similar in the single run through, in the looping example native gets much faster. In a single runthrough, the syscall times are all in the 20-45 us range, but when looped they drop to 4-5 us for all of open/write/read/seek/close.

**Comment by rennergade** [comment]
I've got the Pipe C-module working. The good news is its **much much** faster, the bad news is were still underperforming native. The below graphs are comparing the old deque style pipes to native, and then comparing the new C pipe version to native. You can see that the old version took over an hour to pipe a 1GB file, now with the C pipes it takes 2 seconds. Unfortunately native only takes half a second.

![Screenshot from 2020-12-04 16-36-50](https://user-images.githubusercontent.com/7719272/101261673-069a3400-3707-11eb-9e7b-83ab11470e05.png)
![Screenshot from 2020-12-04 16-36-58](https://user-images.githubusercontent.com/7719272/101261675-08fc8e00-3707-11eb-81af-53b9234fbc21.png)

**Comment by rennergade** [comment]
I also made a flamegraph of the 1GB cat with C-pipes which you can see [here](https://github.com/Lind-Project/lind_project/blob/pipeline-test/graphs/cpipe1g.svg)

What's interesting is that building the Python string after we remove the bytes we want from the pipe takes up so much of read is interesting. Also that the actual pipe writing function "LindPipe_pipewrite" takes up so little of the system call "lind_write" is surprising.

It seems like a decent amount of overhead is the extra copies we have to perform. Still with these cpython pipes we do the following copies

1. Copying from the NaCl buffer into a Python String Object (in nacl)
2. Copying the C-string of the Python String Object into the pipe array (as far as I can tell, PyArg_ParseTuple just exposes the c string pointer within the object, and doesn't do an extra copy)
3. Copying from the pipe array into a buffer in the Read function.
4. Copying that buffer into Py String Object to send back to Python
5. Copying the final buffer back into the NaCl buffer in `lind_read`

I still need to dig into the native kernel pipe code but it's most likely doing at most 2 copies, maybe even just 1.

We can reduce these copies by modifying the code to use c_types strings it seems like. This [StackOverflow post shows ](https://stackoverflow.com/questions/25067790/create-pystring-from-c-character-array-without-copying) that this is possible.

**Comment by rennergade** [comment]
I updated the data I've collected thus far on pipe speeds to include the version with writing to direct buffer addresses, and hacked version where pipe writes and reads don't enter safeposix at all (fake data): you can see [that data here](https://docs.google.com/spreadsheets/d/1G2dtjarFC1lrdw5GdHLDuQYt26L6svnIVMfMnFVkqzQ/edit#gid=0).

I was incorrect in the meeting, the direct buffers does make a slight improvement in speeds from the initial CPipe version, but still not really close to native speed. I also tried limiting direct pipe write/read sizes to 4kB, but that didn't significantly change the speeds here.

The "hacked" version basically hovers around the same speeds as native, a little slower for smaller sizes and faster as the sizes increase. Most of Jonathan's recent fork changes haven't been added to this yet, so there will be some universal gains here and elsewhere, but this hacked version should at least always be faster than native.

I also a generated a flamegraph for the latest pipe implementation which [you can see here.](file:///home/rennergade/developer/nyu/lind/cpipe1g-diraddr.svg)

**Comment by rennergade** [comment]
There's still a bunch of headscratchers here but wanted to document what I've been looking at recently. I've mostly found that Lind still underperforms with I/O in general.

I went back to look at basic I/O using my "filetest1000" example which does: open/write/seek/read/close a thousand times in a loop. Despite equaling performance on some of my other tests, this is a situation where Lind is still slower by quite bit. 

Inspecting this deeper with strace I found that the native kernel for the first ~18 times would do all the system calls in the 20-40 us range. This is still a bit slower than Lind for matching calls which averages about 50% slower. After those 18 times the average system call for native drops down to ~4us, while Lind stays the same, which allows native to be >> faster.

OTOH (intuitively) if I add the O_SYNC flag to the native run while opening the file, Lind becomes much faster than native, which makes me believe that the major factor here is kernel buffering.

Back over to pipes, I did a bunch of reading to figure out if theres any kernel buffering in native specific to pipes. I wasn't able to find anything but I still believe there are some efficiencies here that helps native to outperform us. Native writes/reads for pipes are also those super fast values of 5 us, when I do repeated reads/writes.

The pipe 1G example which as a writer and reader and pipes 1gig of data (a 2^14 buffer 2^16 times), has that average of 5 us per read/write. Oddly, when looking back at the pipepong example, which pipes 1 byte chunks of data, each read and write there averaged 20 us!

I've been trying to do as much reading as I can on I/O but still am left pretty confused, but wanted to get my thoughts on paper.

**Comment by rennergade** [comment]
Over the past several days I've been experimenting with [ftrace](https://www.kernel.org/doc/Documentation/trace/ftrace.txt) to try and get kernel traces to see what is happening to make the above programs so fast natively. I've had some success setting this up but it still need some fine tuning. More specifically, ftrace grabs A LOT of data, and its been hard to filter out background data without losing the information I need to compare. I'll continue to work on this.

**Comment by rennergade** [comment]
I was able to set up Panda and create a QEMU image which I connected to remotely to run GDB with kernel symbols, this helped me get a better sense of what was going on in the kernel and to see if there was caching or buffering that was increasing performance in the filetest for file I/O. This was a pretty great learning experience in digging into the kernel.

I'm still grappling with the results here a bit, and actually went back to check things as I was typing this out. I had run the QEMU image with KVM which gave me fast (~5ms) file syscall times. I had to use the root user to launch it with KVM.  I went back and ran my filetest but non-looped, and again saw the slower (~25ms) syscall times, but when I ran this at root it went down to the 5ms syscall times.

I initially though file I/O was going to be way slower in Lind, and still think it has to be slower (unless we reinvent Lind to use a totally in memory FS open/close/read/write with files still has to do kernel calls, plus we have rpc and book keeping), but how much slower is still a mystery so I'll continue to dig here.

**Comment by rennergade** [comment]
As far as pipes, we don't necessarily need to access the kernel, so we should theoretically be able to approach native speeds (even if they're super fast), so there's hopefully room for improvement here. Brendan mentioned that were still using the kernel for locking and CV's and if we could reduce that we might get better results. I've been looking into lock free queues as a possible template for our pipe system.

**Comment by rennergade** [comment]
I've achieved some awesome initial results over the past week. The LFQ implementation, tested as a pure C program piping 1GB, was faster than native pipes when using large buffer sizes/lower total writes. It was still slower than native when using small buffers/high total writes. 

I used perf/flamegraph to see what was taking more time, and it revealed that it was just doing a ton of mallocs. This was resolved by using [tcmalloc](https://google.github.io/tcmalloc/overview.html) instead of glibc's malloc, which creates large pools for malloc per thread initially and manages them instead of having to do more brk/mmap calls to get memory.

LFQ + tcmalloc outperformed native for all buffer sizes by 40 to 60%. These are huge gains!

Now I'm working on getting these gains into Lind. I'm still having trouble creating the python-pipe shared object file usings distutils.core. I need to figure out how to statically link tcmalloc in or else its overridden by the libc malloc.

The other thing I noticed is that in Lind the RPC times for each call was massively inefficient. I outlined some areas for improvement and passed that to Jonathan who put in a PR that cuts RPC time by 4us per call, which for pipe read/writes is a 33% improvement. He's now working on another fix that should gain us a couple more us for pipe calls.

This will probably take a week or so to get everything fixed and patch the buffer based pipe implementation into the newly improved safeposix framework, plus figuring out how to get the static tcmalloc in the so file, but I am really optimistic that these improvements will help us cross the finish line here, or at least get really close.

**Comment by JustinCappos** [comment]
Can you just statically allocate the buffers and not use malloc altogether?

**Comment by rennergade** [comment]
These buffers are contained in each entry of the LFQ, and basically represent the piped data, plus the LFQ implementation uses malloc for some some things as well. The entrys alone would easily outgrow the stack for most cases.

The tcmalloc approach works really well! Should be able to figure out static linking here.

**Comment by rennergade** [comment]
I was able to create the SO with tcmalloc, and then added my buffer based changes to the improved SafePOSIX pathways. The results are quite good! But still not quite there.

Now to pipe 1G it takes:
Lind - ~600ms
Native - ~470ms

Still not completely there, but getting very close to acceptable margins. Feeling like this is a huge win.

**Comment by rennergade** [comment]
I fixed some efficiency issues and artifacts left in the ResponseBuilder which was slowing this down. Lind time for the above program is down to ~560ms.

**Comment by rennergade** [comment]
Lind is actually now faster than native at piping 1 gigabyte if the buffer size is sufficiently large (so less total writes). Lind is faster than native when buffers are 2^16 bytes or larger. But, as buffers become smaller (more writes) Lind quickly becomes much slower. With 2^4 byte buffers, Lind is taking 10x as long as native.

This is somewhat odd because the pure C test program for the LFQ/TCMalloc implementation trends the opposite way, where the advantage over native becomes greater with smaller buffer sizes. This correlates with our theory that a kernel-less implementation will be faster as more kernel calls have to be made in native.

**Comment by rennergade** [comment]
OTOH, by rearranging how we do some function calls in read and write, the above 1G pipe (2^14 buffers) is down to 515 ms (native 470)

**Comment by rennergade** [comment]
Another thing to note about the baseline native test. I've been running these tests using `/bin/bash` which had root as its owner and group, changing those to the "lind" user/group adds about 20ms for this test, most likely due to further permission checking in the kernel.

**Comment by rennergade** [comment]
Piping 1GB but with the read buffer, as well as the write buffer, reduced to 16 bytes explodes the difference even further. 

Native: 20s
Lind: 770s

**Comment by rennergade** [comment]
A big victory here: `cat` is now faster with Lind than native in all cases, due to its defined buffer sizes

`cat 1gb file | wc` 
Lind: 17.14s
Native: 24.07s
